model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/HotRodMapStorage.java

```diff
@@ -152,29 +190,74 @@ private static String toOrderString(QueryParameters.OrderBy<?> orderBy) {
 
     @Override
     public Stream<V> read(QueryParameters<M> queryParameters) {
+        if (LockObjectsForModification.isEnabled(session, storedEntityDescriptor.getModelTypeClass())) {
+            return pessimisticQueryRead(queryParameters);
+        }
+
+        Query<E> query = prepareQueryWithPrefixAndParameters(null, queryParameters);
+        CloseableIterator<E> iterator = paginateQuery(query, queryParameters.getOffset(),
+                queryParameters.getLimit()).iterator();
+        return closing(StreamSupport.stream(Spliterators.spliteratorUnknownSize(iterator, 0), false))
+                .onClose(iterator::close)
+                .filter(Objects::nonNull) // see https://github.com/keycloak/keycloak/issues/9271
+                .map(this.delegateProducer);
+    }
+
+    private Stream<V> pessimisticQueryRead(QueryParameters<M> queryParameters) {
+        DefaultModelCriteria<M> dmc = queryParameters.getModelCriteriaBuilder();
+
+        // Optimization if the criteria contains only one id
+        String id = (String) dmc.getSingleRestrictionArgument("id");
+
+        if (id != null) {
+            // We have a criteria that contains "id EQ 'some_key'". We can change this to reading only some_key using read method and then apply the rest of criteria.
+            MapModelCriteriaBuilder<K,V,M> mapMcb = dmc.flashToModelCriteriaBuilder(new MapModelCriteriaBuilder<>(keyConverter, fieldPredicates));
+            V entity = read(id);
+            if (entity == null) {
+                return Stream.empty();
+            }
+            boolean fulfillsQueryCriteria = mapMcb.getKeyFilter().test(keyConverter.fromString(id)) && mapMcb.getEntityFilter().test(entity);
+            if (!fulfillsQueryCriteria) {
+                // entity does not fulfill whole criteria, we can release lock now
+                HotRodLocksUtils.removeWithInstanceIdentifier(locksCache, getLockName(id));
+                return Stream.empty();
+            }
+
+            return Stream.of(entity);
+        }
+
+        // Criteria does not contain only one id, we need to read ids non-pessimistically and then read entities one by one pessimistically
+        Query<Object[]> query = prepareQueryWithPrefixAndParameters("SELECT id ", queryParameters);
```

Comment:
During the optimistic locking development, I encountered problems with projection of `id` field for entities that don't have `projectable = true` annotation on `id` field. It was failing with the following error:
`ISPN028502: Unknown alias: authenticatedClientSessions.` for the query `SELECT id FROM kc.HotRodUserSessionEntity c WHERE ((c.realmId = :realmId0) AND (c.offline = :offline0) AND (c.authenticatedClientSessions.clientId = :authenticatedClientSessionsclientId0)) ORDER BY lastSessionRefresh ASC` 
Here it is not failing probably because there is not a pessimistic locking in place for affected entities. Not sure if we want to fix it here or we will leave it for the optimistic locking PR.

Reply:
I would say this is something we should first discuss with the Infinispan team, whether it should work or not. I thought we don't need a projectable index for actually projecting the field, it just makes it faster. Based on your comment it is not the case.

If we will need to add the projectable index to each id that doesn't have it now, it cannot be done in any of the PRs as adding a new index requires migration of all entities. It means we cannot merge any of the PRs until the index is added to each entity in KC 22 and migrated to the new `entityVersion`. Then we can merge for KC 23.

Or, we can skip no-downtime upgradability and suggest reindexing in KC22.